File: investai/backend/models/limite_categoria.py
from .base import ModeloBase, db


class LimiteCategoria(ModeloBase):
    """Limite de gasto mensal que o usuário define para uma categoria
    (ex.: 'lazer' até R$ 300/mês). Usado para o alerta de estouro de
    limite por categoria (RF13)."""

    __tablename__ = "limite_categoria"
    __table_args__ = (
        db.UniqueConstraint("usuario_id", "categoria", name="uq_limite_usuario_categoria"),
    )

    categoria = db.Column(db.String(40), nullable=False)
    valor_limite = db.Column(db.Float, nullable=False)

    usuario_id = db.Column(
        db.Integer, db.ForeignKey("usuario.id"), nullable=False
    )
    usuario = db.relationship("Usuario", back_populates="limites_categoria")

    def atualizar(self, valor_limite=None):
        """UPDATE: altera apenas os campos informados."""
        if valor_limite is not None:
            self.valor_limite = valor_limite
        return self.salvar()

    def to_dict(self):
        return {
            "id": self.id,
            "categoria": self.categoria,
            "valor_limite": self.valor_limite,
            "usuario_id": self.usuario_id,
        }

    def __repr__(self):
        return f"<LimiteCategoria {self.categoria} ({self.valor_limite})>"

File: investai/frontend_flutter/lib/models/limite_categoria.dart
/// RF13 - limite de gasto mensal que o usuário define para uma categoria.
class LimiteCategoria {
  final int? id;
  final String categoria;
  final double valorLimite;

  LimiteCategoria({
    this.id,
    required this.categoria,
    required this.valorLimite,
  });

  factory LimiteCategoria.fromJson(Map<String, dynamic> json) {
    return LimiteCategoria(
      id: json['id'],
      categoria: json['categoria'] ?? '',
      valorLimite: (json['valor_limite'] ?? 0).toDouble(),
    );
  }
}
